### ISSUE TITLE
Localization support

### ISSUE THREAD

**Issue by Ev1lbl0w** [LibreSprite/LibreSprite#49]
LibreSprite is currently lacking a localization system. Since there is already interest from the community to translate this project, we should implement such a system.

We can integrate the popular `i18n` project for this, or create the backend ourselves; which one would you prefer?

After this is implemented, we should also setup a [Weblate](https://weblate.org/en/) or [Crowdin](https://crowdin.com/) *(both free for open-source)*, so everyone can contribute to the localization effort.

**Comment by maujin111** [comment]
Yes we should, @jjhaggar, we need to start planning that too, @Ev1lbl0w what languaje do you like to traslate it?

**Comment by Ev1lbl0w** [comment]
> @Ev1lbl0w what language do you like to translate it?

I would like to translate to Portuguese (Portugal)

But we still have quite some work to do before other people can even start translating stuff. But it's great to see there's already a big interest for this :smiley:

**Comment by jjhaggar** [comment]
No problem @maujin111, as soon as there is something to translate I'm ready to go at it, so please let me know if there is any advances with this issue :)
A question, do you think we should try to make a Neutral Spanish translation or maybe we should make a Latin American Spanish and also a Spain Spanish translation?

**Comment by maujin111** [comment]
@jjhaggar Neutral Spanish is the best way to go i think

**Comment by maujin111** [comment]
we need to vote for a localization system that fits our needs

**Comment by Ev1lbl0w** [comment]
Most of the content lies in the XML files from `data/widgets`. GNU `gettext` can extract strings [from them](https://www.gnu.org/software/gettext/manual/html_node/Preparing-ITS-Rules.html).

I'd prefer to use this tool.

**Comment by maujin111** [comment]
@jjhaggar What do you think?

**Comment by jjhaggar** [comment]
I'm not an expert so I will adapt to the method you all think it's best :)

From my small experience when localizing videogames, I found myself very comfortable when we were using one XML file for each language, and we all edited the translations in a Google Docs spreadsheet (we had a somewhat ugly but automatized process using LibreOffice to pass the strings from Spreadsheet to XML).
![imagen](https://user-images.githubusercontent.com/4078392/86949840-8e2b4900-c14f-11ea-9b26-46ef73b018a7.png)

I have never used gettext, but I've just found there are Windows binaries https://mlocati.github.io/articles/gettext-iconv-windows.html just in case.

From what I have seen in Libresprite repo, there are lots of  XML files in data/widgets Is that ok? When we have 3 languages there would be easily more than 120 files ^^U Is there any possibility of grouping them in one big XML file? Or maybe it doesn't matter? (as said before I'm no expert, so I'm sorry if I say any dumb thing).

**Comment by Ev1lbl0w** [comment]
In my opinion it's just way better to use online platforms for localization, such as [Weblate](https://weblate.org/en/) or [Crowdin](https://crowdin.com/). They offer both intuitive UIs and have way more functionality dedicated to translation than a spreadsheet app.

I worked with CSV/XML formats in the past, and I find them not only very ugly when they contain 20+ languages, but also inefficient for a parser.

The PO format is widely used and supported in open-source software, and a way superior alternative in my opinion. It isn't a pleasant format to edit by hand, but the usage of an online platform would take care of that issue. And for file generation/merging *(which is  a cumbersome process, I'll admit)*, we could setup a CI for that.

> Is there any possibility of grouping them in one big XML file?

There is, and I definitely vote for grouping the files, otherwise it can quickly get cluttered with hundreds of files.

**Comment by maujin111** [comment]
In my personal translation of the program, i created a folder called "es" place there all the xml translated files, modify the source code to load the xml files from that folder. It's not perfect yet, there are still some parts in English but it's a work in progress.

I'll take a look at what @Ev1lbl0w suggested

**Comment by maujin111** [comment]
Putting all xml files in one would require a big change in the source code and a lot of bugs. What do you think @PerryHugh @MintStudios?

**Comment by jjhaggar** [comment]
Yeah, the "ES" folder could also work, I was just giving ideas :)

**Comment by Ev1lbl0w** [comment]
> Putting all xml files in one would require a big change in the source code and a lot of bugs. What do you think @PerryHugh @MintStudios?

I don't think we need to touch the file structure of of `data/widgets`. We will have to alter a lot of text anyways, but I don't see a need to change that file structure.

What I believe we can group together are all the text strings scattered from all files; that can be grouped to one file, and then we code a system to fetch the text from this file.

**Comment by MintStudios** [comment]
@maujin111 I think it depends on the work involved. Changing the source code a lot wouldn't be a good idea. If the transition is controlled and smooth, I think we should, but changing the source code wouldn't be a good idea if the change is huge.

**Comment by PerryHugh** [comment]
Bundling with an external program like gettext isn't what I consider to be ideal, though this would work in a pinch. Preferably the program will not need to access the internet in order to function for alternative languages either, there is no point limiting accessibility to the application's functionality if it can be avoided.

Large scale source code modification is inevitable, strings that need to be localized are spread throughout the source as well as XML files. For example, calls to Alert::show() have hard-coded, specially formatted strings, all in English.

We may want to hook into functions like Widget::setText() and send any string to be displayed to a translation function that compares default English strings to an external file containing the preferred language's strings. We might even be able to do this at a lower level in SHE, but I am not sure yet. Either way, this method is slow, messy, and unmaintainable. It would have to serve as a temporary band-aid.

Essentially, the program was not designed for localization. English is scattered throughout the repositories in all sorts of places. We might have to, at least initially, hook into the processing of text and defer to a custom solution for translating.

**Comment by jjhaggar** [comment]
> I believe we can group together are all the text strings scattered from all files; that can be grouped to one file, and then we code a system to fetch the text from this file.

I think that would be a great help for facilitating the localization process. There's probably lots of people that would love to help adding their own languages, and maybe they could feel intimidated of using software or web services that are foreign to them, but almost everybody is able to open one plain XML file with their text editor of choice and start translating strings :)

I know that what I've just described is probably not the best practice, but I think it would be good to have it also as an option for potential translators.

**Comment by Zughy** [comment]
If this is still on the table, I'd like to contribute with Italian translation. For my experience, Minetest is using Weblate and it's a very smooth process. They usually give a 2 weeks period of feature freeze where translators are encouraged to translate in their own language and, just before releasing, they execute "some sort of script" to import translations (I don't know how exactly how this last part works)
